mesh: Change API to deliver tokens via JoinComplete

przemyslawfierek-silvair opened the pull request:
This patch changes Import and CreateNetwork API, in current implementation mentioned functions don't return token. Token is passed to application via the JoinComplete method call. When application doesn't raise any error during handling JoinComplete then it is assumed that the token has been saved, otherwise when application replies with an error message then the Node is removed.

michallowasrzechonek-silvair committed:
mesh: Add mesh-io for Silvair Radio over UART or TCP

michallowasrzechonek-silvair committed:
mesh: Check that incoming packet fits in net key cache

michallowasrzechonek-silvair committed:
mesh: Replace aes_cmac with aes_cmac_one

michallowasrzechonek-silvair committed:
mesh: Add option to use openssl for crypto

jakubwitowski-silvair committed:
mesh: Add DBus method for updating sequence number

przemyslawfierek-silvair committed:
mesh: Add D-Bus API to 'Access Control List' management to the 'mesh-io-tcpserver'

This patch provides D-Bus API to access control management. From now it
is possible to grant or revoke access for 'uGW' using single D-Bus call.

przemyslawfierek-silvair committed:
mesh: Add D-Bus initialization API to 'mesh-io'

This patch provides API to perform D-Bus initialization to 'mesh-io'.
This should allow to create some D-Bus objects related to 'mesh-io'.

michallowasrzechonek-silvair committed:
mesh: Fix FPE in overcommit logic.

During overcommit, mesh_config_save is called in asynchronous mode to
avoid blocking Send() calls. This means that update of cfg->write_time
is scheduled via l_idle_oneshot, so if the next Send() gets scheduled
first, the code may see elapsed time of zero.

If this happens, then the overcommit logic was already executed and the
overcommit is pending, so we can just return.

michallowasrzechonek-silvair committed:
mesh: Compatilibty fix for openssl-1.0.2

michallowasrzechonek-silvair commented on mesh/mesh.c:
```diff
@@ -430,11 +430,18 @@ static void send_join_failed(const char *owner, const char *path,
 {
 	struct l_dbus *dbus = dbus_get_bus();
-	struct l_dbus_message *msg;
+	struct l_dbus_message *method_call;
```
Brian is not going to like it, let's leave it as `msg` to keep the patch as small as possible. Or name it `call` if you must.

michallowasrzechonek-silvair commented on mesh/mesh.c:
```diff
@@ -445,7 +452,7 @@ static bool prov_complete_cb(void *user_data, uint8_t status,
 
 	owner = join_pending->sender;
-	path = join_pending->app_path;
+	path = node_get_app_path(join_pending->node);
```
This is worth a separate commit, I think - it fixes a genuine bug in the daemon.

michallowasrzechonek-silvair replied:
Also, is it possible to remove `app_path` from `join_pending` altogether?

michallowasrzechonek-silvair commented on mesh/mesh.c:
```diff
@@ -662,11 +667,21 @@ static struct l_dbus_message *leave_call(struct l_dbus *dbus,
 {
 	struct l_dbus_message *reply;
 	struct l_dbus_message *pending_msg;
+	struct l_dbus_message *method_call;
```
I would rename it to just `call`.

michallowasrzechonek-silvair commented on mesh/mesh.c:
```diff
@@ -675,20 +690,29 @@ static void create_node_ready_cb(void *user_data, int status,
 
-	l_debug();
-	l_dbus_message_set_arguments(reply, "t", l_get_be64(token));
+	node_attach_io(node, mesh.io);
```
I think this should happen only after `JoinComplete` succeeds.

michallowasrzechonek-silvair commented on mesh/mesh.c:
```diff
@@ -675,20 +690,29 @@ static void create_node_ready_cb(void *user_data, int status,
-	}
 
-	node_attach_io(node, mesh.io);
+		l_dbus_send(dbus_get_bus(), reply);
```
I would cache `dbus_get_bus` in a local variable.

michallowasrzechonek-silvair reviewed (changes requested):
API documentation needs to be updated, and doc changes should come *first* in the patchset.

przemyslawfierek-silvair committed:
mesh: Fix invalid app_path on 'Join'

This patch fixes invalid app_path on 'Join' method call.

przemyslawfierek-silvair committed:
mesh: Change API to deliver tokens via JoinComplete

This patch changes Import and CreateNetwork API, in current
implementation mentioned functions don't return token. Token is passed
to application via the JoinComplete method call. When application doesn't
raise any error during handling JoinComplete then it is assumed that
the token has been saved, otherwise when application replies with
an error message then the Node is removed.

michallowasrzechonek-silvair commented:
Nice.

przemyslawfierek-silvair committed:
tools: Add to 'mesh-cfgclient' waiting for 'JoinComplete'

michallowasrzechonek-silvair commented:
Merged upstream, thanks.
